### PR TITLE
Supermatter Explosion Releases an EMP

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

When the supermatter explodes, an EMP will be released with a minimum size of 40,60

## Why It's Good For The Game

More stuff to happen when the supermatter explodes, the entire station will be taken offline for a short amount of time and the power grid will actually be damaged. Will drain the power from batteries, so power generation from solars will need to be set up.
Currently the SM explosion is just a tiny little explosion which is usually of the minimum size doing almost nothing.

IPCs will fear the supermatter and will either have to run as far away from the SM as they can get, or stay somewhere where they can get repaired. Technology will be in ruins and the station will be back to the stone age.

## Testing Photographs and Procedure

image

## Changelog
:cl:
add: The supermatter explosion will now create a relatively large EMP.
/:cl:

### CODE REVIEW

**Comment by Rukofamicom** [comment]
Given the rework to how EMP affects everything, I think this is great, but I would go so far as to say it should at least light EMP the entire Z level so the effects aren't localized in engineering. 

I still overall prefer the idea in https://github.com/BeeStation/BeeStation-Hornet/pull/6857 with tweaks I suggested there. 

Supermatter going ded should have a station-wide effect, but not necessarily an irreversible one. Just a very inconvenient and/or annoying one that makes sabotage a viable option without incompetence resulting in a ruined/prematurely ended round. The scale of the destructive explosion borders on irreversible in many cases given its scale. I don't even think I need to mention Tesla/Singularity.

**Comment by williamcll** [comment]
Yeah I like the anomaly SM more. if you want an engine that EMPs the station go port Storm drive.

**Comment by PowerfulBacon** [comment]
Well I intend this to work alongside the anomaly PR, rather than having 1 or the other

**Comment by williamcll** [comment]
> Well I intend this to work alongside the anomaly PR, rather than having 1 or the other

I think that's a bit too much unless your intentions are to force shuttle call like almost /vg/

**Comment by Mat05usz** [comment]
> I think that's a bit too much unless your intentions are to force shuttle call like almost /vg/

Yeah both happening at the same time (with current intensity) will most certainly force lots of rounds to end early. If you intended on having both, either of them has to be less intense

**Comment by Rukofamicom** [comment]
@Mat05usz @williamcll keep in mind that EMPs are a lot less intense as of recently.
Station-wide EMP is more like a stationwide blackout + damage to silicons and IPCs. A few wires will get pulsed at random, but nothing will actually be totally destroyed now. 

https://github.com/BeeStation/BeeStation-Hornet/pull/6135

> EMP no longer drain energy weapons, only disabling them temporarily
All machines are now temporarily disabled by EMPs instead of potentially destroyed

**Comment by williamcll** [comment]
> @Mat05usz @williamcll keep in mind that EMPs are a lot less intense as of recently.
Station-wide EMP is more like a stationwide blackout + damage to silicons and IPCs. A few wires will get pulsed at random, but nothing will actually be totally destroyed now.

Not the issue here. You can fuck silicons hard by blowing the SM now. Especially on malf AI which now must keep engineering locked out as well as their own sat.

**Comment by PowerfulBacon** [comment]
You also fuck over the entire station by blowing SM. I like the idea of having something as a trade off. (EMP only disables silicons for a short amount of time and drains some power which they can get back from a charger, AI is basically unaffected except for the SMES losing some charge, camers going offline temporarilly (which it can fix) and the turrets disabling for some time.)

**Comment by williamcll** [comment]
> You also fuck over the entire station by blowing SM. I like the idea of having something as a trade off. (EMP only disables silicons for a short amount of time and drains some power which they can get back from a charger, AI is basically unaffected except for the SMES losing some charge, camers going offline temporarilly (which it can fix) and the turrets disabling for some time.)

No it forces them into APC mode since EMPs have a high chance to turning off the main breaker as well as draining APC power.

**Comment by ghost** [comment]
> Given the rework to how EMP affects everything, I think this is great, but I would go so far as to say it should at least light EMP the entire Z level so the effects aren't localized in engineering.

EMPs to IPCs are relatively difficult to repair, and on a station struggling for power, IPCs are even LESS likely to get fixed up. Unless this has a REALLY weak strength I see a station-wide EMP as a mostly unavoidable giant fuck you to any IPC, which is far worse than one you can avoid.

**Comment by Rukofamicom** [comment]
I don't see any issues with a racial weakness actually being relevant semi-regularly tbh.

File: code/game/objects/empulse.dm [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/proc/empulse(turf/epicenter, heavy_range, light_range, log=0, magic=FALSE, holy=FALSE)
+/proc/empulse(turf/epicenter, heavy_range, light_range, log=0, magic=FALSE, holy=FALSE, check_tick=FALSE)
```
Comment (Archanial) [inline code/game/objects/empulse.dm:1]: Why does it need a flag to check ticks

File: code/modules/power/supermatter/supermatter.dm [hunk 2]
```diff
@@ -301,6 +302,11 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 			L.rad_act(rads)
 
 	var/turf/T = get_turf(src)
+
+	//Big empulse
+	INVOKE_ASYNC(GLOBAL_PROC, /proc/empulse, T, 200 * max(0.2, gasmix_power_ratio), 300 * max(0.2, gasmix_power_ratio), TRUE, FALSE, FALSE, TRUE)
```
Comment (Archanial) [inline code/modules/power/supermatter/supermatter.dm:307]: How big is it?

**Comment by williamcll** [comment]
> I don't see any issues with a racial weakness actually being relevant semi-regularly tbh.

Because this is being stacked with the /tg/ SM anomalies PR.

File: code/game/objects/empulse.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/game/objects/empulse.dm:1]: Because the emp is so big it causes lag

File: code/modules/power/supermatter/supermatter.dm (hunk 2 above)
Comment (PowerfulBacon) [reply code/modules/power/supermatter/supermatter.dm:307]: Most of the station usually

**Comment by PowerfulBacon** [comment]
This is cool but too much effort to resolve conflicts on a PR that will be open for another 2 months anyway

**Pull request closed without merging**